File: data structures/C++/findDuplicates.cpp
// O(N)
// O(1)
#include <vector>
#include <map>

class Solution {
public:
    std::vector<int> findDuplicates(std::vector<int>& nums) {

        if(nums.size()== 0) return std::vector<int>{};

        std::vector<int> result;
        std::map<int,int> map;

        for(int i = 0; i < nums.size(); i++)
        {
            if(map.find(nums[i])!= map.end())
                result.push_back(nums[i]);
            else
            map[nums[i]]++;
        }

        return result; 
    }
};

File: data structures/C#/FindDuplicates.cs
//O(n)
//O(n)
public class Solution
{
    public IList<int> FindDuplicates(int[] nums)
    {

        if (nums.Length < 0) return new List<int> { };

        var dict = new Dictionary<int, int>();
        var duplicateNumbersList = new List<int>();

        for (int i = 0; i < nums.Length; i++)
        {
            if (dict.ContainsKey(nums[i]))
                duplicateNumbersList.Add(nums[i]);
            else
                dict.Add(nums[i], i);
        }

        return duplicateNumbersList;
    }
}